lib/data/dpa/dpClient.js

```diff
@@ -429,12 +461,154 @@ class DpClient {
+                }
+                log.debug('Recovered fragments. Re-writing..',
+                    { lostFragsIds });
+                const recoveredFrags = allFrags.filter((frag, idx) =>
+                    lostFragsIds.indexOf(idx) > -1);
+
+                let noError = true;
+                let count = 0;
+                // re-write lost fragments
+                recoveredFrags.forEach((frag, idx) => {
+                    const fragKey = keys[lostFragsIds[idx]];
+                    const stream = createReadbleStream(frag, fragSize);
+                    client.put(stream, fragSize, null, key, err => {
+                        if (err) {
+                            log.error('error from datastore',
+                                { error: err, implName });
+                            if (noError) {
+                                noError = false;
+                                return cb(err);
+                            }
+                        }
+                        count++;
+                        if (noError && count === recoveredFrags.length) {
+                            log.debug('Recovered frags are stored',
+                                { lostFragsIds });
+                            return cb();
+                        }
+                        return undefined;
+                    }, fragKey);
+                });
+                return undefined;
+            });
+            return undefined;
+        }
+
+        keys.forEach((keyFrag, fragIdx) => {
+            this.checkAvailFrag.bind(this)(keyFrag, log, (err, val) => {
+                if (err) {
+                    lostFragsIds.push(fragIdx);
+                } else {
+                    fragments.push(val);
+                }
+                if (lostFragsIds.length + fragments.length === codeLen) {
+                    if (lostFragsIds.length > 0) {
+                        return repairFrags.bind(this)(lostFragsIds, fragments,
+                            callback);
+                    }
                     return callback();
                 }
                 return undefined;
             });
         });
     }
+
+    /**
+     * This checks availability of a fragment
+     * @param {String} key - The key associated to the fragment
+     * @param {object} log - Logger
+     * @param {Callback} callback - callback (err, fragment)
+     * @returns {undefined}
+     */
+    checkAvailFrag(key, log, callback) {
+        const maxRetriesNb = this.cronParams.retriesNb;
+        const interval = this.cronParams.interval;
+
+        assert.strictEqual(typeof key, 'string');
+
+        function getFrag(key, reqUids, retriesNb, cb) {
+            log.debug('Checking availability fragment',
+                { key, retry: retriesNb });
+            client.get(key, null, reqUids, (err, val) => {
+                if (err) {
+                    const newRetriesNb = retriesNb + 1;
+                    if (newRetriesNb > maxRetriesNb) {
```

Comment:
This is not important but you begin to 0, and then check if it's more than maxRetriesNb witch is 3.
So the first fail you will increase to 1, then to 2, and again to 3.
3 is not more than 3 so you will try it again, and then return the cb with the error, 
At this end it's 4 shots.